src/polymer/custom-style-extension.ts

```diff
@@ -0,0 +1,128 @@
+
+/**
+ * @license
+ * Copyright (c) 2018 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import * as dom5 from 'dom5';
+import {Document, isPositionInsideRange, ParsedHtmlDocument, Replacement, Severity, Warning} from 'polymer-analyzer';
+
+import {HtmlRule} from '../html/rule';
+import {getIndentationInside} from '../html/util';
+import {registry} from '../registry';
+import {indentSourceRange, stripIndentation} from '../util';
+
+const p = dom5.predicates;
+const isCustomStyleV1 = p.AND(
+    (node: dom5.Node) =>
+        !node.parentNode || !p.hasTagName('custom-style')(node.parentNode),
+    p.hasTagName('style'),
+    p.hasAttrValue('is', 'custom-style'));
+
+class CustomStyleExtension extends HtmlRule {
+  code = 'custom-style-extension';
+  description = stripIndentation(`
+      Warns when \`custom-style\` is used as type extension on the main document.
+      This:
+          <style is="custom-style">
+            div {
+              --app-color: red;
+            }
+          </style>
+
+      Should instead be written as:
+          <custom-style>
+            <style is="custom-style">
+              div {
+                --app-color: red;
+              }
+            </style>
+          </custom-style>
+  `);
+
+  async checkDocument(parsedDocument: ParsedHtmlDocument, _document: Document) {
+    const warnings: Warning[] = [];
+
+    const customStyleTags = dom5.queryAll(
```

Comment:
New dom5 api returns an iterable, but we can get an array easily enough out of it:

```typescript
const customStyleTags = [...dom5.queryAll(parsedDocument.ast, isCustomStyleV1, dom5.childNodesIncludeTemplate)];
```